presto-password-authenticators/src/main/java/io/prestosql/plugin/password/file/FileBasedAuthenticator.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.password.file;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.airlift.log.Logger;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.security.AccessDeniedException;
+import io.prestosql.spi.security.BasicPrincipal;
+import io.prestosql.spi.security.PasswordAuthenticator;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.security.Principal;
+import java.util.Map;
+
+import static io.prestosql.plugin.password.file.Users.readUsers;
+import static io.prestosql.spi.StandardErrorCode.CONFIGURATION_INVALID;
+
+public class FileBasedAuthenticator
+        implements PasswordAuthenticator
+{
+    private static final Logger log = Logger.get(FileBasedAuthenticator.class);
+    private Cache<PasswordStore, Boolean> bcryptCache;
+    private Cache<PasswordStore, Boolean> pbkdf2Cache;
+    private Map<String, User> authMap;
+
+    @Inject
+    public FileBasedAuthenticator(FileBasedAuthenticationConfig config)
+    {
+        bcryptCache = CacheBuilder.newBuilder().concurrencyLevel(1)
+                .maximumSize(config.getAuthTokenCacheStoreMaxSize()).build();
+
+        pbkdf2Cache = CacheBuilder.newBuilder().concurrencyLevel(1)
+                .maximumSize(config.getAuthTokenCacheStoreMaxSize())
+                .build();
+        try {
+            this.authMap = readUsers(config.getConfigFile(), config.getBcryptMinCost(), config.getPbkdf2MinIterations(), bcryptCache, pbkdf2Cache);
+        }
+        catch (IOException e) {
+            throw new PrestoException(CONFIGURATION_INVALID, "Failed to read password database: " + config.getConfigFile(), e);
+        }
+    }
+
+    @Override
+    public Principal createAuthenticatedPrincipal(String inputUsername, String inputPassword)
+    {
+        // username does not exist in the database
+        if (!this.authMap.containsKey(inputUsername)) {
```

Comment:
Nit: no need for `this` here

Also, let's avoid the double map lookup
```java
User user = authMap.get(inputUsername);
if (user == null || !user.authenticate(inputUsername, inputPassword)) {
    throw new AccessDeniedException("Invalid credentials");
}

Reply:
Done